File: codes/semana12/Ex02_Transporte.py
from abc import ABC, abstractmethod

# Produto abstrato
class Transporte(ABC):

    @abstractmethod
    def viajar(self):
        pass


# Produtos concretos
class Onibus(Transporte):

    def viajar(self):
        print("Ônibus realizando rota urbana.")


class Taxi(Transporte):

    def viajar(self):
        print("Táxi transportando passageiro.")


class Bicicleta(Transporte):

    def viajar(self):
        print("Bicicleta compartilhada iniciando trajeto.")


# Factory Method
class FabricaTransporte:

    @staticmethod
    def criar_transporte(tipo):

        if tipo == "onibus":
            return Onibus()

        elif tipo == "taxi":
            return Taxi()

        elif tipo == "bicicleta":
            return Bicicleta()

        else:
            raise ValueError("Tipo de transporte inválido")


# Programa principal
tipo = input("Escolha o transporte: ")

transporte = FabricaTransporte.criar_transporte(tipo)

transporte.viajar()

File: codes/semana12/Ex02_Transporte.cpp
#include <iostream>
#include <memory>

using namespace std;

// Produto abstrato
class Transporte {
public:
    virtual void viajar() = 0;
    virtual ~Transporte() {}
};

// Produtos concretos
class Onibus : public Transporte {
public:
    void viajar() override {
        cout << "Ônibus realizando rota urbana." << endl;
    }
};

class Taxi : public Transporte {
public:
    void viajar() override {
        cout << "Táxi transportando passageiro." << endl;
    }
};

class Bicicleta : public Transporte {
public:
    void viajar() override {
        cout << "Bicicleta compartilhada iniciando trajeto." << endl;
    }
};

// Factory
class FabricaTransporte {
public:

    static unique_ptr<Transporte> criar(string tipo) {

        if (tipo == "onibus")
            return make_unique<Onibus>();

        else if (tipo == "taxi")
            return make_unique<Taxi>();

        else if (tipo == "bicicleta")
            return make_unique<Bicicleta>();

        return nullptr;
    }
};

int main() {

    string tipo;

    cout << "Escolha o transporte: ";
    cin >> tipo;

    auto transporte = FabricaTransporte::criar(tipo);

    if (transporte)
        transporte->viajar();
    else
        cout << "Tipo de transporte invalido!" << endl;

    return 0;
}